design/secure-credentials-store.md

```diff
@@ -21,51 +21,119 @@ Credentials include:
- plain-text in etcd, but for a production system this is not recommended as such Kubernetes secrets may be a good option
- for deployments on Kubernetes/OpenShift as there would be a single place to protect.
- 
- HashiCorp Vault - a popular open-source vault.  
- 
- 
- ## Should everything be stored in a vault?
- 
- Client credentials, user credentials and external tokens may not be suitable to store in a vault due to the high number
- of such items. It may be better to encrypt these within the database.
- 
- 
- # Per-type configuration?
- 
- It may be beneficial to allow per-type configuration. For example retrieving LDAP and SMTP credentials from Kubernetes secrets, while using HashiCorp Vault Transit Secrets for external tokens and not encrypting user secrets like otp secrets. That would allow more flexiblity as well as tuning security vs performance.
+The task will be split into two separate parts:
+
+1. Read-only access to a secure storage (i.e. vault). Vault is managed and written-to by external
+   tools specific to vault implementation.
+
+2. Encryption / decryption SPI for secrets. There are services providing encryption / decryption
+   of secrets, while abstracting from key management and other stuff.
+
+The rest of this document will deal with the first part (Vault SPI) only. **The second part
+(encryption / decryption) will be treated in a separate document.**
+
+
+## Scope
+
+Generally any secret can be stored in a vault. The value stored in Keycloak database should
+however be only a reference to the vault provided by the administrator.
+
+For some types like user OTP secret or client credentials it might be impractical to store
+them into a vault: to change a secret, the user would have to change it in the vault. Yet the
+decision what is practical and what not needs to be kept in administrator's hands. Hence any
+place where a secret is retrieved should support a way to obtain it from the vault that needs to be
+consistent across the whole product.
+
+Generally, there might be more than one vault SPI provider available at the same time.
+For example, when a new vault is plugged in addition to an existing one, the values
+from the original vault are likely not be migrated immediately, and thus the references
+to the original vault should work.
+
+### Realm Keys
+
+Today there are SPIs both for realm keys as well as signing of tokens. Keys SPI enables storing keys in Java KeyStore as well as adding custom support for a vault. Signature SPI enables delegating token signature to external HSM devices.
+
+## Design
+
+### Vault Provider SPI
+
+It may be beneficial to allow per-type configuration. For example retrieving LDAP and SMTP credentials from Kubernetes secrets, while using HashiCorp Vault Transit Secrets for external tokens and not encrypting user secrets like otp secrets. That would allow more flexiblity as well as tuning security vs performance.
```

Comment:
Apparently my yesterday's reply to this comment has been lost, sorry about that.

In short - I believe we should not. Vault (`VaultProvider`) is an aggregation of named vault service providers (`VaultServiceProvider`). `VaultServiceProvider` is an implementation of a specific vault, thus it should not be aware of the its vault ID, which is only an id of the instance within `VaultProvider`.

For example, some secrets (e.g. LDAP, realm private keys) might need to be stored very securely (expensive HSM with small capacity), the rest of secrets might be stored elsewhere (software vault). As an administrator, one can specify the secrets for given purpose (realm keys, LDAP) should be retrieved from specific vault(s) instead of hardcoding the vault ID in a property name. Furthermore, keycloak realm administrator need not even be aware of the existence of two or more different vault types.

Purpose maps to several vaults, which allows easy transition from one vault to another (e.g. from software to HSM), if the secret IDs can stay the same (e.g. `${vault.realm-key}`)

Reply:
> In short - I believe we should not. Vault (VaultProvider) is an aggregation of named vault service providers (VaultServiceProvider). VaultServiceProvider is an implementation of a specific vault, thus it should not be aware of the its vault ID, which is only an id of the instance within VaultProvider.

@hmlnarik Thanks. This clarifies it a bit. I though the hierarchy is the other way around. Could we put that explanation somewhere in the design?

> Furthermore, keycloak realm administrator need not even be aware of the existence of two or more different vault types.

@hmlnarik I've always assumed, that Keycloak Admin is the same persona that configures Keycloak though the admin console and configures the deployment (e.g. configuring the SPIs in `standalone.xml`). 

> Purpose maps to several vaults, which allows easy transition from one vault to another (e.g. from software to HSM), if the secret IDs can stay the same

@hmlnarik Ok, thanks for the explanation

Reply:
Let's keep things simple then. There will be no purpose and there will be only a single vault per server for the initial implementation. If there would be a need for more, we can either add purpose into Keycloak context, or extend the interface later with default implementation.